### PR TITLE
Implement normal broadcasting semantics with grid=False

### PR DESCRIPTION
**Pull request by nsmith-**
Although good for making 2D grids of the pdf value, if we want to evaluate `mypdf.xfxQ` on a per-sample basis, it is more convenient to have the function broadcast, e.g.
```python
nnpdf31 = parton.mkPDF("NNPDF31_lo_as_0118", 0)
nnpdf31.xfxQ(
    np.array([1, 21, 21]),
    np.array([0.211, 0.038, 0.0581]),
    np.array([877, 231, 1.6e+03]),
    grid=False,
)
```
returns `array([0.17802354, 2.65219091, 1.44782397])` with this patch, whereas previously it would error because the x and Q values are not monotone increasing (as they would be for e.g. calls to `np.meshgrid`). With `grid=True` (the default) it should behave as before.

I would argue one is better off treating this function like a [ufunc](https://numpy.org/doc/stable/reference/ufuncs.html) and require the user to construct the grid with `np.meshgrid` for cases as in the original usage. This would be an API change though.

### CODE REVIEW

**Comment by DavidMStraub** [comment]
Thanks for the contribution! @peterstangl would you have time to review?

**Comment by rappoccio** [comment]
FYI I cross-checked this against `lhapdf` native API and it gives the same results

```
    nnpdf31 = lhapdf.mkPDF("NNPDF31_lo_as_0118", 0)                                                                                                                                                                                 
    print(nnpdf31.xfxQ2(1, 0.427, 1.64e+03))                                                                                                                                                                                        
    print(nnpdf31.xfxQ(1, 0.427, np.sqrt(1.64e+03)))                                                                                                                                                                                
                                                                                                                                                                                                                                    
    nnpdf31_again = parton.mkPDF("NNPDF31_lo_as_0118", 0)                                                                                                                                                                           
    print(nnpdf31_again.xfxQ(np.array([1]), np.array([0.427]), np.sqrt(np.array([1.64e+03])), grid=False) )  
```

**Comment by peterstangl** [comment]
> Thanks for the contribution! @peterstangl would you have time to review?

I would have time to review at the end of this week / beginning of next week.

**Comment by garvitaa** [comment]
Any updates on this?

**Comment by peterstangl** [comment]
I thought I'd have time for a review this week, but unfortunately some very important things have come up, so I won't get to it until the end of next week.

File: parton/pdf.py [hunk 1]
```diff
@@ -45,10 +45,16 @@ def __call__(self, x, y, *args, **kwargs):
...
+            if kwargs.get("grid", True):
+                if any_out_of_bounds_x:
+                    z[out_of_bounds_x, :] = self.fill_value
+                if any_out_of_bounds_y:
+                    z[:, out_of_bounds_y] = self.fill_value
+            else:
+                if any_out_of_bounds_x:
+                    z[out_of_bounds_x] = self.fill_value
+                if any_out_of_bounds_y:
+                    z[out_of_bounds_y] = self.fill_value
```
Comment (peterstangl) [inline parton/pdf.py]: This does not work in general, in particular if `x` and `y` have different shape and are broadcasted against each other (see https://github.com/scipy/scipy/blob/v1.10.0/scipy/interpolate/_fitpack2.py#L930-L1008). I think an easy fix would be to add
```python
if not kwargs.get("grid", True) and x.shape != y.shape:
    x, y = np.broadcast_arrays(x, y)
```
before `out_of_bounds_x` and `out_of_bounds_y` are defined.

In addition, if `kwargs['grid'] == False`, it would be enough to modify `z` once :
```python
if any_out_of_bounds_x or any_out_of_bounds_y:
    z[out_of_bounds_x | out_of_bounds_y] = self.fill_value
```

File: parton/pdf.py [hunk 2]
```diff
@@ -162,10 +168,17 @@ def interpolator(self, flavor):
...
         """Return x*f(x) for flavor `flavor`, momentum fraction `x` and
         squared factorization scale in units of GeV^2, `Q2`."""
-        return self.interpolator(flavor)(np.log(x), np.log(Q2))
+        flavors = np.unique(flavor)
+        if grid and len(flavors) > 1:
+            raise RuntimeError("No logical way to make a grid for multiple flavors")
+        out = self.interpolator(flavors[0])(np.log(x), np.log(Q2), grid=grid)
+        for f in flavors[1:]:
+            mask = flavor == f
+            out[mask] = self.interpolator(f)(np.log(x), np.log(Q2), grid=grid)[mask]
```
Comment (peterstangl) [inline parton/pdf.py]: This implementation implicitly assumes that `flavor`, `x`, and `Q2` all have the same shape. However, when implementing broadcasting, I think one should only assume that `flavor`, `x`, and `Q2` are broadcastable to the same shape.

File: parton/pdf.py [hunk 3]
```diff
@@ -182,25 +195,23 @@ def __init__(self, name, member=0, pdfdir=None):
...
-    def xfxQ2(self, flavor, x, Q2):
+    def xfxQ2(self, flavor, x, Q2, grid=True):
         """Return x*f(x) by specifying flavor, `x`, and factorization scale
         squared `Q2` in GeV^2."""
-        res = np.empty((np.size(x), np.size(Q2)))
-        res[:] = np.nan
+        if grid == True:
+            res = np.full((np.size(x), np.size(Q2)), np.nan)
+        else:
+            res = np.full(np.size(x), np.nan)
```
Comment (peterstangl) [inline parton/pdf.py]: This implicitly assumes that `flavor`, `x`, and `Q2` all have the same shape (and also that they are 1D), which should not be a requirement if broadcasting is used.

File: parton/pdf.py [hunk 4]
```diff
@@ -182,25 +195,23 @@ def __init__(self, name, member=0, pdfdir=None):
...
         for i, pdfgrid in enumerate(self.pdfgrids):
-            _res = np.atleast_2d(pdfgrid.xfxQ2(flavor, x, Q2))
+            _res = pdfgrid.xfxQ2(flavor, x, Q2, grid=grid)
```
Comment (peterstangl) [inline parton/pdf.py:222]: Why did you remove the `np.atleast_2d`? Is it clear that this will never be needed?

**Review by peterstangl — CHANGES_REQUESTED** [review]

File: parton/pdf.py (hunk 4 above)
Comment (nsmith-) [reply parton/pdf.py:222]: When `grid=True` then the output of scipy.interpolate.RectBivariateSpline is always 2D regardless of the shape of the inputs. When `grid=False` anyway the shape will be whatever `np.broadcast_shapes(flavor, x, Q2)` is.

**Comment by nsmith-** [comment]
Thanks for the review! Sorry for the delay in implementing the changes. I also added a test

**Comment by peterstangl** [comment]
I think this PR looks very good now and should be ready to be merged.

**Comment by peterstangl** [comment]
The only thing to add might be a requirement on the numpy version in `setup.py`. The PR uses `broadcast_shapes`, which was introduced in numpy v1.20.0.

**Comment by nsmith-** [comment]
Hmm, numpy v1.20.0 was released Jan 30, 2021, is that old enough that we're ok with the requirement? Otherwise I could just use `broadcast_arrays` and `full_like`

**Comment by peterstangl** [comment]
Since numpy versions prior to 1.20 are no longer officially supported anyway, I think it should be safe to have this requirement. I would just suggest to add it to `setup.py`.

**Comment by nsmith-** [comment]
Ah, great! Indeed https://numpy.org/neps/nep-0029-deprecation_policy.html#support-table means we don't need to support it. I'll bump the requirement.

**Comment by nsmith-** [comment]
Anything further preventing this from merge?

**Comment by DavidMStraub** [comment]
Thanks for the nudge, looks good!
